Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp

```diff
@@ -0,0 +1,403 @@
+        }
+#endif
+
+        for (const auto& button : profile.buttons) {
+            if (!isBindingForHand(button.hand))
+                continue;
+
+            const auto& actions = m_buttonActions.get(button.type);
+            if (button.flags & OpenXRButtonFlags::Click) {
+                ASSERT(actions.press != XR_NULL_HANDLE);
+                CHECK_XRCMD(createBinding(profile.path, actions.press, makeString(m_subactionPathName, button.path, s_pathActionClick), bindings));
+            }
+            if (button.flags & OpenXRButtonFlags::Touch) {
+                ASSERT(actions.touch != XR_NULL_HANDLE);
+                CHECK_XRCMD(createBinding(profile.path, actions.touch, makeString(m_subactionPathName, button.path, s_pathActionTouch), bindings));
+            }
+            if (button.flags & OpenXRButtonFlags::Value) {
+                ASSERT(actions.value != XR_NULL_HANDLE);
+                CHECK_XRCMD(createBinding(profile.path, actions.value, makeString(m_subactionPathName, button.path, s_pathActionValue), bindings));
+            }
+        }
+
+        for (const auto& axis : profile.axes) {
+            auto action = m_axisActions.get(axis.type);
+            ASSERT(action != XR_NULL_HANDLE);
+            CHECK_XRCMD(createBinding(profile.path, action, makeString(m_subactionPathName, unsafeSpan(axis.path)), bindings));
+        }
+    }
+
+    return XR_SUCCESS;
+}
+
+std::optional<PlatformXR::FrameData::InputSource> OpenXRInputSource::collectInputSource(XrSpace localSpace, const XrFrameState& frameState) const
+{
+    PlatformXR::FrameData::InputSource data;
+    data.handedness = m_handedness;
+    data.handle = m_handle;
+    data.targetRayMode = PlatformXR::XRTargetRayMode::TrackedPointer;
+    data.profiles = m_profiles;
+
+    getPose(m_pointerSpace, localSpace, frameState, data.pointerOrigin);
+    PlatformXR::FrameData::InputSourcePose gripPose;
+    if (XR_SUCCEEDED(getPose(m_gripSpace, localSpace, frameState, gripPose)))
+        data.gripOrigin = gripPose;
+
+    Vector<std::optional<PlatformXR::FrameData::InputSourceButton>, openXRButtonTypes.size()> buttons;
+    for (auto& type : openXRButtonTypes) {
+        if (auto button = collectButton(type); button.has_value())
+            buttons.append(button);
+    }
+
+    // Trigger is mandatory in xr-standard mapping.
+    if (buttons.isEmpty() || !buttons.first().has_value())
+        return std::nullopt;
+
+    for (size_t i = 0; i < buttons.size(); ++i) {
+        if (buttons[i]) {
+            data.buttons.append(*buttons[i]);
+            continue;
+        }
+        // Add placeholder if there are more valid buttons in the list.
+        for (size_t j = i + 1; j < buttons.size(); ++j) {
+            if (buttons[j]) {
+                data.buttons.append({ });
+                break;
+            }
+        }
+    }
+
+    Vector<std::optional<XrVector2f>, openXRAxisTypes.size()> axes;
+    for (auto& type : openXRAxisTypes)
```

Comment:
here I would again make it by value

```suggestion
    for (auto type : openXRAxisTypes)
```

as there's no benefit in const-ref, and we don't seem to need ref